src/poetry/core/constraints/generic/constraint.py

```diff
@@ -43,21 +67,59 @@ def operator(self) -> str:
         return self._operator
 
     def allows(self, other: BaseConstraint) -> bool:
-        if not isinstance(other, Constraint) or other.operator != "==":
+        """Logic table to help
+
+                ||   !=   |   ==   |   in   | not in
+        --------||--------|--------|--------|--------
+           !=   ||   !=   |   !=   | not in | not in
+           ==   ||   !=   |   ==   |   in   | not in
+           in   || not in |   in   |   in   | not in
+         not in || not in | not in | not in | false
+
+        """
+
+        if not isinstance(other, Constraint) or other.operator not in {
+            "==",
+            "in",
+            "not in",
+        }:
```

Comment:
For `allows()`, `other` must be an exact value, in other words, the operator must be `==`. This will not change with `in` and `not in`.

I assume you can move part of your logic to `allows_all()` and `allows_any()` were `other` can have any operator. Further, you have to replace some `allows()` in `difference()`, `intersect()` and `union()` with `allows_all()` or `allows_any()`.

Reply:
I can revert `allows()` to only consider `other.operator == "=="`, but I have some questions related how you would like me to implement `allows_any()` and `allows_all()` inside `difference()`, `instersect()` and `union()`.

I have searched the code base for uses of `allows_any()` and `allows_all()`, but I have found no uses outside of unit testing. I would like some clarification on usage, as well as confirmation of logic on `intersect()` and `union()` for some operators.

First some examples
|                        |                        |    Intersect      |       Union        |
|----------------|---------------|-----------------|----------------|
| `"tegra" not in`| `"tegra" not in`| `"tegra" not in` | `"tegra" not in` |
| `"tegra" in`      | `"tegra" not in`| empty set         |  all/any set       |
| `"tegra" in`      | `"tegra" in`      | `"tegra" in`        | `"tegra" in`       |
| `"teg" in`         | `"tegra" in`      | `"tegra" in`        | `"teg" in`          |
| `"teg" not in`   | `"tegra" in`      | empty set         | empty set                                               |
| `"tegra" in`      | `"rpi" in`            | empty set       | MultiConstraint( `"tegra" in`,`"rpi" in` )      |
| `!=1.2.3-tegra`   | `"tegra" in`    | `!=1.2.3-tegra`    | MultiConstraint( `!=1.2.3-tegra` ,`"tegra" in` )      |
| `!=1.2.3-tegra`   | `"tegra" not in`    | empty set         |  empty set        |

Reply:
I edited your table directly because I think/hope it's easier to keep track. First of all, this is a finicky topic so I might have made mistakes. If you think that some cell is wrong, please tell me. I crossed out mistakes instead of just replacing them, so you can follow the changes. If you agree feel free to remove the crossed out text.

Some explanations:

- I added a line with `"tegra" in` and `"teg" in` because order is relevant for `allows_all` in this case. (`"teg" in` allows all `"tegra" in`, but `"tegra" in` does not allow all `"teg" in`)
- I think you mixed up `MultiConstraint` and `UnionConstraint` a bit. (`MultiConstraint` is like `and` (intersection) and `UnionConstraint` is like `or` (union).)

Some general rules (I hope they are correct 😉):
- If `c1.allows_all(c2)` is `True`, `c1.allows_any(c2)` must also be `True`.
- If `c1.allows_any(c2)` is `False`, `c1.allows_all(c2)` must also be `False`.
- If `c1.allows_any(c2)` is `True`, `c2.allows_any(c1)` must also be `True`.
- If the intersection is empty, `allows_any` is `False` (for both orders). If `allows_any` is `False` for both orders, the intersection is empty.
- If the intersection is not empty, `allows_any` is `True` (for both orders). If `allows_any` is `True` for both orders, the intersection is not empty.

Reply:
After reviewing and the code, I think there might be some ambiguity around 

|  constraint 1  |  constraint 2 |    Intersect      |       Union        |   allow_any |       allow_all  |
|----------------|---------------|-----------------|----------------|-----------------|----------------| 
| `!=1.2.3-tegra`   | `"tegra" in`    |   MultiConstraint( `!=1.2.3-tegra` ,`"tegra" in` )   |   all/any set    | `True` | `False` |

I think the union should be `UnionConstraint` instead of `AnyConstraint`

Reply:
Can you give an example that is not allowed by `!=1.2.3-tegra or "tegra" in`? (If the union allows all possible values, then it is `AnyConstraint`.)

Reply:
I cannot, I think I was able to fix all the cases above